### ISSUE TITLE
RFC: `unitary` subroutines, as a gate-like subroutine acting on quantum registers and with more versatile control flow

### ISSUE THREAD

**Issue by ndebeaudrap** [Qiskit/openqasm#323]
### Summary

In the "Circuit Families & Generics" workgroup, we have drafted a specification for a variety of subroutine, which we propose to call `unitary` subroutines . These are in a sense intermediate between a `gate` subroutine and a `def` subroutine. The following is a high-level description of such subroutines, and the syntax with which they would be defined and invoked. We propose this as a feature to be added to OpenQASM3.

### Motivation

It is useful to be able to define subroutines which act on aggregate quantum data, grouped into quantum registers. This may be done in principle in OpenQASM3 using def subroutines, but doing so does not allow possibility of inverting these subroutines or controlling them coherently, even when the subroutine happens to consist entirely of unitary operations. Such 'circuit'-level unitary subroutines are commonplace in the quantum literature. Enabling such subroutines to be easily expressed will serve to increase the versatility of OpenQASM3 for programmers, and also for educators. This is also a preliminary step towards making it possible to define portable families of such operations (such as the quantum Fourier transform on varying register sizes) through subroutine declarations.

The distinct features which we aim to capture with `unitary` subroutines, are:
- in contrast to `gate` subroutines, they may take registers as quantum operands;
- also in contrast to `gate` subroutines, the subroutine definition may involve control flow using run-time constants;
- in contrast to `def` subroutines, these subroutines may be inverted and coherently controlled using gate modifiers `inv @` and `ctrl @` respectively.

The intent is to provide a means of defining unitary subroutines of greater versatility than is possible with `gate` subroutines, while retaining some of the advantages of a subroutine in which there is a clear separation between classical program and quantum data (e.g., no measurement or feedback loops between the two kinds of data).

A prototypical typical example would be a quantum Fourier transform, or similarly involved subroutine which one might like to use while also taking advantage of the fact that it is a unitary operation.

These subroutines would in principle be easily compiled to sequences of basic blocks similar to a top-level quantum program, even when invoked with modifiers such as `ctrl @` or `inv @`.

### Proposed syntax, with examples

We present the proposed syntax below, with simple 'toy' examples which may be more typical of an educational setting rather than application.

#### A simple unitary subroutine declaration

The proposed declaration syntax for `unitary` subroutines differs somewhat from that of gates, in that they may have quantum operands consisting of registers of any specified width (or arrays of qubits of some specified shape). Because the types of the operands cannot be assumed to be individual qubits, they must be declared explicitly. (In order to separate the subroutine name from the types of its quantum operands, we introduce a `:` as a delimiter between them.) The unitary operation then acts on those operands, without use of broadcast semantics. For example:

```
    // define a cyclic shift operation on
    // a three-dimensional subspace spanned
    // by |00>, |01>, |10> of a two-qubit state,
    // representing addition of one (mod 3) on
    // an encoded qutrit
    unitary qutritX : qubit[2] a {
      x a[1];
      cx a[1], a[0];
      cx a[0], a[1];
    }

    // define a ‘clock’ operator on a three-dimensional
    // subspace spanned by |00>, |01>, |10>, a two-qubit state
    unitary qutritZ : qubit[2] a {
      rz(2*pi/3) a[0];
      rz(-2*pi/3) a[1];
    }
```

In both cases the operand is a single register `a` consisting of two qubits. (When invoked, the subroutines `qutritX` and `qutritZ` may act on a 'register' which is in fact a slice of some larger register.)

As the syntax may suggest, `unitary` subroutines may not involve any explicitly non-unitary operations such as `reset` or `measure`. Nor may they be defined using any subroutines apart from `gate` subrotuines, and other `unitary` subroutines.

#### Unitary subroutine invocations, with and without gate modifiers

In addition to acting on quantum registers as operands, `unitary` subroutines may be controlled or inverted, using the same modifiers allowed for `gate` subroutines:

```
    // define a controlled-addition gate on two qutrits,
    // where each qutrit is encoded in two qubits
    unitary qutritCX : qubit[2] c, qubit[2] t {
      ctrl @ qutritX c[0], t;
      ctrl @ inv @ qutritX c[1], t;
    }
```

In this example, we define a `unitary` subroutine acting on two operands, each of which is a register of two qubits. The unitaries are defined by modifications of the `qutritX` subroutine, in each case using the `ctrl @` modifier, which adds a single qubit to the beginning of the list of quantum operands, as with `gate` subroutines. Using these declarations, we may then invoke `qutritCX` on the top level of an OpenQASM3 program as follows:

```
    qubit[2] a;
    qubit[2] b;
    h a[0];
    qutritCX a, b;
    qutritZ a;
```
Note that the invocations of `qutritCX` and `qutritZ` are in effect similar in style to invocations of gates. However, they do not involve any broadcast semantics. In particular, invocations of these subroutines are only well-typed if they are invoked on quantum registers whose widths match the operands described in the respective subroutine declarations.

#### Classical parameters and control flow in unitary subroutines

In common with `gate` subroutines, a `unitary` subroutine may also be defined using zero, one, or more classical parameters. Also in common with `gate` subroutines, these parameters are specified in parentheses, before all of the quantum operands. However, in contrast to `gate` subroutines, these parameters are not assumed to have type `angle`, and so their types should be explicitly declared. Also in contrast to `gate` subroutines, these classical arguments may be used for control flow statements in the body of the subroutine. For example:
```
    // define an operation which simulates adding an
    // integer (mod 3) to the state of a qutrit,
    // encoded in two qubits
    unitary qutritPlus (int b) : qubit[2] a {
      if (b % 3 == 1) qutritX a;
      else if (b % 3 == 2) inv @ qutritX a;
    }
```
The control flow allowed in the declaration of a `unitary` subroutine cannot involve any measurements, and therefore cannot involve any feedback between the classical parameters and the quantum operands. In order to allow such a subroutine to be coherently controlled and inverted, we treat all parameters in the subroutine as being immutable. It would also be allowed to define local variables within the scope of a `unitary` subroutine, as a convenience to the programmer, but these too would be immutable once initialised. This behaviour extends also to the iterators of `for` loops, which may be used in the definition of a `unitary` subroutine:
```
    unitary QFT64 : qubit[6] q {
      uint n = 6;
      for j in [0 : n-1] {
        h q[j];
        for k in [j+1 : n-1]
          ctrl @ Rz(pi / 2**(k-j+1)) q[j], q[k];
        }
      for uint j in [0:n-1]
        if (j != n-1-j)
          swap q[j], q[n-1-j]
    }
```
While the iterators `j` and `k` above change their values in the course of their loops, we treat them as run-time constants. This allows the inverse or a coherently controlled version of this same circuit to be easily computable from the subroutine definition. Thus, using the above subroutine `qutritPlus`, one may write a subroutine invocation as follows:
```
    input int s;
    qubit[2] t;
    qubit c;

    H c;

    ctrl @ qutritPlus(s) c, t;
```
Note that the additional quantum operand (a single qubit) introduced by the ctrl @ modifier, occurs after the classical argument `s` to the subroutine qutritPlus, as it would for a controlled `gate` invocation.

### Applications and future possibilities

The proposed feature of `unitary` subroutines stand on their own, in principle, as a feature which may be adopted for OpenQASM3. This feature may be expected to be most useful to those who wish to write circuits with medium-to-large scale repeated structures, or who wish to write multiple programs involving a common medium-to-large scale subroutine. A simple example would be defining non-trivial penalty functions in terms of boolean formulae on a register of more than a handful of qubits.

It is anticipated that `unitary` subroutines will be most useful in combination with a capacity to define 'generic' subroutines: this feature is to be proposed separately, and would be applicable both to `unitary` subroutines and `def` subroutines. In particular, it would make it possible to define quantum Fourier transforms as a _family_ of unitary subroutines acting on quantum operands of various sizes, without the need to re-implement it for each operand size required in practise.

**Comment by ajavadia** [comment]
I chatted with @stevenheidel about this proposal. It seems like this proposal is basically a restriction on subroutines, i.e. a subroutine built out of unitary components and control-flow that is statically resolvable. So my main question is whether we can leverage most of what the subroutine syntax already gives us, rather than define a new language construct. In particular, can we "tag" a unitary subroutine as a special case of subroutine, something like:

```
unitary def qutritPlus (int b) : qubit[2] a {
      if (b % 3 == 1) qutritX a;
      else if (b % 3 == 2) inv @ qutritX a;
    }
```

The compiler should raise an error if there is any usage of `measure` or `reset` in the body, or if there is an assignment to `b`. Otherwise we would be free to invoke `ctrl` and `inverse` on `unitary subroutine` types.

(Side note: initially I thought about extending `gate`, but I decided against that. Gate declarations in QASM are for defining unitary matrices (hierarchically from a basic universal set of `u`/`ctrl`/`gphase`). The body is not prescriptive for how that gate is implemented. In contrast this proposal is much closer to subroutines where there is a prescription of the instructions that should happen to implement that subroutine)

Last question: when you say classical variables have to be immutable, doesn't that mean they should be defined as `const`? That seems like a good way of enforcing this.

**Comment by ndebeaudrap** [comment]
My main concern about regarding these as `def` subroutines, can be summarised by saying that I don't believe that the syntax for declaring (and similarly, invoking) `def` subroutines is as you describe it above. Aren't all of the quantum operands of a `def` subroutine, passed along with the other classical parameters?

The issue with this is: if you then apply a `ctrl @` modifier to this subroutine, you have to decide where to put the additional quantum operand. One could simply add it to the front of the list of parameters, but then the syntax for how `ctrl @` modifies the invocation of `unitary def` subroutines would be distinct from the syntax for how it modifies the invocation of `gate` subroutines. In my opinion at least, it would be desirable to maintain consistency. (Also: as with `gate` subroutines, the fact that the classical parameters are immutable motivates regarding the quantum operands as being quite different from the classical parameters. This motivates separating the classical parameters from the quantum operands, again as one does for `gate` subroutines.)

It is possible that, 'under the hood', the compiler might think of a `unitary` subroutine as a special variety of `def` subroutine: that the way that it is represented in intermediate stages is a minor elaboration of that case. I see no obvious obstacles to that, provided that the `unitary` subroutine body is consistent with the constraints described above on `unitary` declarations. If treating them this way makes them easier to implement, so much the better. However, the similarities to the way that `gate` subroutines are used motivates having a different syntax from `def` subroutines, unless we aim to extend or modify the syntax of `def` subroutines in general.

**Comment by ndebeaudrap** [comment]
Regarding declaring classical variables as `const`: I see an argument for requiring the programmer to declare each of them explicitly as `const`, especially if in the end it is decided to extend/modify the `def` subroutine syntax to accomodate `unitary` subroutines (or modify the proposed syntax for `unitary` subroutines to fit the existing `def` syntax).

Also: even if `unitary` subroutines are defined with a distinct syntax from `def` subroutines, I can still see an argument for requiring the programmer to explicitly indicate that classical variables are `const`, though in this case I think it would be less urgent. There would be a trade-off between requiring the programmer to do the bookwork of explicitly applying a type modifier which is in fact mandatory (making them jump through hoops to a small extent), versus the clarity of having the cosntraints of the variable being explicitly reinforced (even if from context it should be very easy for the programmer to infer those constraints). I think this would warrant some discussion with the Types and Casting WG.

**Comment by meamy** [comment]
Re: the question of `gate` versus `def`, can't `unitary` declarations be viewed as defining a family of unitary matrices indexed by classical parameters? This makes the feature strictly more powerful than `unitary def`s. Without generics the `gate` version effectively allows the definition of larger `gate`s than are reasonably possible with the `OPENQASM 2.0` `gate` syntax, something I think the `QFT` example points out.

My hunch is that a `gate`-style `unitary` should also be cleaner to spec formally in syntax, immutable classical parameters notwithstanding.

**Comment by stevenheidel** [comment]
In my conversation with Ali, one thing we agreed on is that "gate declarations in QASM are for defining unitary matrices".

Let's assume that we disallow conditionals within a gate declaration (more on that later). Something like "QFT256gate" can still be represented as a unitary matrix. But, as you point out, it leads to a lot of typing in both the declaration and the invocation.

**Declaration**

On declaration, the issue that we have is two-fold:
1. "The arguments in qargs cannot be indexed within the body of the gate definition."
2. For loops are not allowed within the body of a gate statement

Assuming we allowed both, the QFT256gate could be written as follows (with no local variable definitions and a strawman syntax):
```
gate QFT256 q[8] {
    for j in [0:8] {
        h q[j];
        for k in [j+1:8] {
            ctrl @ Rz(pi / 2**(k-j+1)) q[j], q[k];
        }
    }
    for j in [0:4]
        swap q[j], q[7-j];
}
```

This would be equivalent the the ~40 line statement listed above. At this point, we'd just be introducing a new _shorthand_ syntax for defining large gates. The idea of the gate, and its possible representation as a unitary matrix, are still valid here.

**Invocation**

In the qutrit example, we have the gate:

```
gate qutritX a[2] {
    x a[1];
    cx a[1], a[0];
    cx a[0], a[1];
}
```

But when you try to call it, you have to unpack each qutrit:

```
qubit[2] myQutrit;
qutritX myQutrit[0] myQutrit[1];
```

This is because in OpenQASM "for convenience, gates automatically broadcast over registers." To solve this issue, we'd need a syntax to turn off the broadcasting and send each qubit in a register to a separate qubit in the list of qargs. (Perhaps the unpacking syntax from Python: `qutritX *myQutrit`.

**On conditionals**

I'm not convinced by the above example that conditionals are worth adding to gate definitions, and breaking the representation as a unitary matrix. The above example could be two gates: qutritAddOne and qutritAddTwo.

**Summary**

In summary, I believe we have three high level strategies to tackled this proposal:
1. Extend the gate declaration and invocation to support shorthand syntax (what I'm advocating for above)
2. Treat these as a subset of subroutines (`unitary def`), but deal with the issue at invocation of the `ctrl` modifier
3. Introduce a new structure (this proposal)

I'm very hesitant to do 3, because it makes it much harder to process gate invocations. Previously when we encountered a gate invocation that's all it could be, but now we'll need to determine whether it's instead a `unitary`, which completely changes the semantics.

**Comment by ndebeaudrap** [comment]
Thanks, Steven, I have a few thoughts on your proposal.

**Regarding the issues with 'Strategy 3'**

I would like to understand a bit more about the issues underlying what you call 'Strategy 3' for implementing more complex unitary subroutines.

Suppose you have an invocation of a subroutine, which could be a `gate` or could be a distinct `unitary` subroutine. It is clear that, at present, anything which has a 'gate-like' invocation (with separate classical prameters and quantum operands) must be a `gate`, and that 'Strategy 3' would change this, as `unitary` subroutines would also have gate-like invocation. So it is clear that some additional work would have to be done to determine the operation to be performed.

How much additional work would this involve? Naively, I would expect that upon reading a 'gate-like' invocation, one might consult a symbol table in any case to see whether the subroutine being invoked has been defined. One piece of information that the symbol table could include, is whether the identifier is a `gate` or a different sort of subroutine. Having identified that it is a `gate, the compiler could presumably carry on as it currently does; having identified that it is a different sort of subroutine, it does something else. This information regarding whether the identifier is a `gate` or something else, could be recorded when the subroutine has been defined.

So: assuming that the above is correct, the question appears to boil down, to how much added work is required to modify the structure of the symbol table. Do I understand correctly that this is likely to be quite a lot of work? Is it possible to explain what the issues are there? Or is there some other problem that I have missed above?

**Regarding 'Strategy 1' in general**

I have no objections, in general, with regarding so-called `unitary` subroutines as 'a way of specifying unitary matrices', in the sense that the compiler does not necessarily have to promise to implement the operation by performing the operations described by the programmer. (I assume here that the compiler would not necessarily attempt to compute an explicit matrix, but would use the declaration as a guide to obtaining _some_ suitable representation for the matrix.)

_Regarding 'if' statements_

It's not clear to me why it is possible to accomodate `for` loops as a way of extending `gate` definitions, but not `if` statements. If there is any intention of allowing `gate` definitions to have arguments which might not be angles, then a `for` loop may have a bound which depends on an input parameter; an `if` statement could be treated similarly. In either case, depending on the value of some parameter (or perhaps some compile-time expression), an equivalent OpenQASM2-style `gate` definition could be determined at compile time and put in place.

Of course, if the intent is not to allow `gate` subroutines to have parameters which are not angles, then this is a functionality where the proposed functionality extends beyond what `gate` subroutines can support. It would also likely lead to problems when we consider how this functionality could be extended by allowing meta-parameters (defining not just a single unitary operator, but a family of unitary operators — such as the QFT on registers of various sizes).

Is there some difficulty involving `if` statements, or in allowing non-angle parameters, that I am missing?

_Regarding broadcasting_

If we provisionally consider the possibility that `gate` subroutines could be defined with _some_ control flow (such as `for` loops, whether or not we allow `if` statements), then the main issue I see with interpreting these subroutines as `gate`s is precisely the issue of broadcasting. It is indeed possible that we could overcome this by defining syntax for 'non-broadcastable' registers. (Indeed, this is an issue that the Types and Casting WG mused about some months ago in the context of `def` subroutines and arrays, but to no definite conclusion.)

I think that defining a special syntax for 'non-broadcastable registers', with 'broadcastable registers' being the default, is undesirable. Naturally I think that it is important to maintain backward compatibility with OpenQASM2, regarding broadcasting, at least for the time being. But requiring an explicit definition or indication that a register is 'non-broadcasting', is likely to arise frequently when we use these more complex unitary subroutines; and in particular we would have to address how this would be done when the 'register' is actually a slice or an alias. I think that this is likely to lead to cruft in the way OpenQASM is written, in the long run, and so I think this approach should be avoided. I think that this motivates giving 'Strategy 3' serious consideration, as a means of representing complicated unitary transformations which have different semantics when invoked on registers.

**Summary**

I appreciate that there may have to be some significant changes to how the compiler works, to accomodate this functionality. I would like to understand better why the things you ywould like to avoid, would be difficult to implement.

It is concievable that we can implement some interesing subset of my proposal, by trying to interpret these unitary subroutines as an extension of `gate` subroutines. My concern is that in doing so, we will end up making OpenQASM3 more complicated for programmers to write in practise, and furthermore make it difficult to support any of the other desirable features of this proposal. As this proposal is ultimately about a forward-looking feature (which will become more important with time), it seems important to me to make sure that we don't paint ourselves into a corner in how it is implemented.

**Comment by stevenheidel** [comment]
> How much additional work would this involve? Naively, I would expect that upon reading a 'gate-like' invocation, one might consult a symbol table in any case to see whether the subroutine being invoked has been defined.

This is exactly how it would be implemented and it is not that much additional work for a reasonably sophisticated compiler. However we would lose an advantage that we have right now: that gate calls can be distinguished at parse time prior to doing any further compiler passes. The gate calls resolve to a [separate AST node](https://github.com/Qiskit/openqasm/blob/main/source/openqasm/openqasm3/ast.py#L523). This makes it very simple to manipulate programs involving lists of gate invocations. I would guesstimate that 95%+ of OpenQASM programs "in the wild" are merely a list of gate invocations, followed by measurement statements.

My other concern with around invocation looking the same between gates and unitaries is that one call is broadcasted and the other isn't. We had the same problem with subroutines while writing the paper. It would be unclear to the reader whether registers were broadcast or not. This is why we moved qubit arguments, which were originally outside the parenthesis like gates, to inside the parenthesis.

> I have no objections, in general, with regarding so-called unitary subroutines as 'a way of specifying unitary matrices'
> Is there some difficulty involving if statements, or in allowing non-angle parameters, that I am missing?

You're not missing anything involving if statements. It's really a question of runtime vs compile-time resolvable parameters.

Imagine we only allow angle parameters, and all gates must resolve to a unitary matrix at compile-time. Then there would still be a use for an if statement. For instance, while iterating through a loop variable I imagine the statement `if i % 2 == 0` could be useful for performing a gate every other time. Therefore ignore my opposition to if statements in general.

For strategy 1 to work, the problem is not with if statements but rather with non-angle parameters. If non-angle parameters are introduced, then gates would become "a family of unitary matrices indexed by classical parameters" rather than a unitary matrix, as they are now. Strategy 1 in my mind relies on keeping the purpose of gates the same, while adding some additional syntax to add structure and reduce repetition.

> It would also likely lead to problems when we consider how this functionality could be extended by allowing meta-parameters (defining not just a single unitary operator, but a family of unitary operators — such as the QFT on registers of various sizes).

This is what I'm interesting in learning more about tomorrow. I'd like to explore whether we can restrict classical parameters to only this purpose (defining sizes of registers) and how that would look for both strategies 1 and 3.

**Comment by stevenheidel** [comment]
I reviewed the revised proposal and support it (up to some minor details). I would like to ensure a majority of the TSC supports the high-level direction of the revised proposal before we jump into the finer points.

@ajavadia @bettinaheim  @blakejohnson @levbishop @pschindler

**Comment by blakejohnson** [comment]
I support the proposal.
